File: Python3/프로그래머스/3/43164. 여행경로/여행경로.py
import collections
import sys

def solution(tickets):
    sys.setrecursionlimit(10000)
    
    graph = collections.defaultdict(list)
    total_cnt = collections.defaultdict(int)
    for ticket in tickets:
        graph[ticket[0]].append(ticket[1])
        total_cnt[(ticket[0], ticket[1])] += 1
        
    def backtracking(ans, visited):
        if len(ans) == len(tickets) + 1:
            result.append(ans[:])
            return
        
        cur = ans[-1]
        for next in graph[cur]:
            if (cur, next) not in visited or visited[(cur, next)] < total_cnt[(cur,next)]:
                ans.append(next)
                visited[(cur, next)] += 1
                backtracking(ans, visited)
                ans.pop()
                visited[(cur, next)] -= 1
           
    result = []
    visited = collections.defaultdict(int)
    backtracking(["ICN"], visited)
    result.sort()
    return result[0]

File: Swift/프로그래머스/3/43164. 여행경로/여행경로.swift
import Foundation

func solution(_ tickets:[[String]]) -> [String] {
    
    var result = [String]()
    let total = tickets.count
    
    // 1. 그래프 생성
    var graph = [String: [String]]()
    for ticket in tickets {
        graph[ticket[0], default: []].append(ticket[1])
    }
    
    // 우선순위 정렬하기
    for (k,v) in graph {
        graph[k] = graph[k]!.sorted()
    }
    
    var flag = false
    
    func backtracking(_ cur_country: String, _ ans: [String]) {
        if flag {
            return
        }
        
        if ans.count == total + 1 {
            result = ans
            flag = true
            return
        }
            
        let country_list = graph[cur_country, default: []]
        if country_list.count == 0 { return }
        
//        print(ans, "            ", graph[cur_country])
        
        var temp = [String]()
        temp = country_list
        
        for (i, next_country) in temp.enumerated() {
//            print(i, temp)
            var next_country = temp[i]
            temp.remove(at: i)
            graph[cur_country] = temp
            
            backtracking(next_country, ans + [next_country])
            
            temp = country_list
            graph[cur_country] = temp
        }
    }
    
    backtracking("ICN", ["ICN"])
    
    return result
}